src/Generators/Microsoft.Gen.Logging/Emission/Emitter.Method.cs

```diff
@@ -276,132 +165,317 @@ static string ConvertPropertyToString(LoggingProperty lp, string arg)
+                    {
+                        numClassifiedProps++;
+                    }
+                    else
                     {
-                        break;
+                        numUnclassifiedProps++;
                     }
+                }
 
-                    index++;
+                if (p.HasProperties && !p.SkipNullProperties)
+                {
+                    p.TraverseParameterPropertiesTransitively((_, member) =>
+                    {
+                        if (member.ClassificationAttributeType != null)
+                        {
+                            numClassifiedProps++;
+                        }
+                        else
+                        {
+                            numUnclassifiedProps++;
+                        }
+                    });
                 }
+            }
 
-                // check for an index that's too big, this can happen in some cases of malformed input
-                if (index < templateParameters.Length)
+            if (!string.IsNullOrEmpty(lm.Message))
+            {
+                numUnclassifiedProps++;
+            }
+
+            numReservedUnclassifiedProps = numUnclassifiedProps;
+            numReservedClassifiedProps = numClassifiedProps;
+
+            if (numReservedUnclassifiedProps > 0)
+            {
+                OutLn();
+                OutLn($"_ = {stateName}.ReservePropertySpace({numReservedUnclassifiedProps});");
+                int count = numReservedUnclassifiedProps;
+                foreach (var p in lm.Parameters)
                 {
-                    var parameter = templateParameters[index];
-                    var atSign = parameter.NeedsAtSign ? "@" : string.Empty;
-                    if (parameter.PotentiallyNull)
+                    if (NeedsASlot(p) && p.ClassificationAttributeType == null)
                     {
-                        const string Null = "\"(null)\"";
-                        OutLn($"var {atSign}{t.Key} = {lambdaHelperName}[{holderMap[parameter]}].Value ?? {Null};");
-                        result = true;
+                        var key = $"\"{lm.GetParameterNameInTemplate(p)}\"";
+
+                        if (p.IsEnumerable)
+                        {
+                            var value = p.PotentiallyNull
+                                ? $"{p.NameWithAt} != null ? {LoggerMessageHelperType}.Stringify({p.NameWithAt}) : null"
+                                : $"{LoggerMessageHelperType}.Stringify({p.NameWithAt})";
+
+                            OutLn($"{stateName}.PropertyArray[{--count}] = new({key}, {value});");
+                        }
+                        else
+                        {
+                            var value = ShouldStringify(p.Type)
+                                ? ConvertToString(p, p.NameWithAt)
+                                : p.NameWithAt;
+
+                            OutLn($"{stateName}.PropertyArray[{--count}] = new({key}, {value});");
```

Comment:
Can be extracted outside of `if ... else` block

Reply:
Done.